File: backend/app/models/order.py
from app.models.product import ProductVariant
from pydantic import BaseModel, EmailStr
from typing import Optional
from datetime import datetime
from prisma.enums import PaymentMethod, ShippingMethod, OrderStatus, PaymentStatus
from app.models.user import User
from app.models.address import Address

class OrderItemLite(BaseModel):
    id: int
    name: Optional[str]
    variant_id: Optional[int]
    variant: Optional[ProductVariant]
    price: float
    quantity: int
    image: Optional[str]

class OrderItemCreate(BaseModel):
    variant_id: int
    quantity: int
    price: float

class OrderCreate(BaseModel):
    status: Optional[OrderStatus] = OrderStatus.PENDING
    payment_status: Optional[PaymentStatus] = PaymentStatus.PENDING
    coupon_id: Optional[int] = None

class Order(BaseModel):
    id: int
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    order_number: str
    user_id: int
    user: Optional[User]
    shipping_address: Optional[Address]
    total: float
    subtotal: float
    tax: float
    discount_amount: float = 0
    wallet_used: Optional[float] = 0
    status: OrderStatus
    payment_status: PaymentStatus
    shipping_method: ShippingMethod
    payment_method: PaymentMethod
    shipping_fee: float
    coupon_code: Optional[str] = None
    order_items: Optional[list[OrderItemLite]]
    order_notes: Optional[str] = None
    invoice_url: Optional[str] = None
    created_at: datetime

class PaginatedOrders(BaseModel):
    items: list[Order]
    next_cursor: int | None
    limit: int

class OrderTimelineEntry(BaseModel):
    id: int
    order_id: int
    from_status: Optional[OrderStatus] = None
    to_status: Optional[OrderStatus] = None
    message: Optional[str] = None
    created_at: datetime

File: app/src/schemas/order.ts
import { z } from "zod";
import { OrderStatusSchema, ShippingMethodSchema, PaymentMethodSchema, PaymentStatusSchema } from "./enums";
import { UserSchema } from "./user";
import { AddressSchema } from "./address";
import { ProductVariantLiteSchema } from "./product";
import { CursorSchema } from "./common";

export const OrderItemSchema = z
    .object({
        id: z.number(),
        name: z.string(),
        variant: ProductVariantLiteSchema,
        image: z.string().optional(),
        quantity: z.number(),
        price: z.number(),
    });

export const OrderSchema = z
    .object({
        id: z.number(),
        email: z.string(),
        phone: z.string().optional(),
        order_number: z.string(),
        user: UserSchema,
        shipping_address: AddressSchema,
        total: z.number(),
        subtotal: z.number(),
        tax: z.number(),
        discount_amount: z.number(),
        wallet_used: z.number(),
        status: OrderStatusSchema,
        payment_status: PaymentStatusSchema,
        shipping_method: ShippingMethodSchema,
        payment_method: PaymentMethodSchema,
        shipping_fee: z.number(),
        coupon_code: z.string().optional(),
        order_items: z.array(OrderItemSchema),
        order_notes: z.string().optional(),
        invoice_url: z.string().optional(),
        created_at: z.string(),
    });

export const PaginatedOrdersSchema = CursorSchema.extend({
    items: z.array(OrderSchema),
});

export type Order = z.infer<typeof OrderSchema>;
export type OrderItem = z.infer<typeof OrderItemSchema>;
export type PaginatedOrders = z.infer<typeof PaginatedOrdersSchema>;
